rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+}
+
+fn main(){
+   App::new()
+   .add_plugins(DefaultPlugins)
+   .add_system(tick_construction_timer)
+   .add_system(
+      update_construction_progress
+        .after(tick_construction_timer)
+         // We can add systems with read-only system parameters as run criteria
+        .run_if(construction_timer_finished)
+   )
+   .add_system(system_meltdown_mitigation.run_if(too_many_enemies))
+   // We can use closures for simple one-off run criteria, 
+   // which automatically fetch the appropriate data from the `World`
+   .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+   // The `resource_exists` and `resource_equals` functions generate a closure that you can stick in a run criteria
+   .add_system(gravity.run_if(resource_exists(Gravity)))
+   // Run criteria can be attached to system sets, allowing them to control all systems in that set
+   .configure_set(GameSet::Physics.run_if(resource_equals(Paused(false))))
+   .run();
+}
+```
+
+There are a few important subtleties to bear in mind when working with run criteria:
+
+- when multiple run criteria are attached to the same system, the system will run if and only if all of those run criteria return true
+- each run criteria are evaluated once during each pass of the schedule
+  - this occurs just before the first system controlled by that criteria would be run
+  - the value is then shared for all other systems with that run criteria, ensuring all systems in the set stay synchronized
+
+This is important to ensure that the state of the world always matches the state expected by the system at the time it is executed.
+
+#### States
+
+**States** allow you to toggle systems on-and-off based on the value of a given resource and smoothly handle transitions between states by running cleanup and initialization systems.
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three groups of systems:
+
+1. **While-active:** these systems run each schedule iteration if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(apply_damage.run_if(state_equals(GameState::Playing))`
+2. **On-enter systems:** these systems run once when the specified state is entered.
+   1. `app.add_system(generate_map.in_set(OnEnter(GameState::Playing))`
+3. **On-exit systems:** these systems run once when the specified state is exited.
+   1. `app.add_system(autosave.in_set(OnExit(GameState::Playing)))`
+
+While-active systems are by far the simplest: they're simply powered by run criteria.
+
+On-enter and on-exit systems are stored in dedicated schedules, two per state, within the `App's` `Schedules`.

```

Comment:
The state transition handler needs to be an exclusive system because the "on-enter" and "on exit" sets (and any set in general) could contain exclusive systems (like command flushes). While that isn't the ultimate ideal for performance, it's consistent behavior, so I'd consider that good UX.

The system executor keeps pre-allocated and cached data structures to be efficient, so we aren't gonna be able to literally splice system sets into a currently running schedule without significantly complicating the executor internals.

Your best bet for "on-demand execution" will be commands, which can be used exactly like exclusive systems (the `&mut World` is the important part). I consider that accounted for here, and I don't think we can reasonably do better than that.